refactor for 2.0.0

TimTaylor opened the pull request:
This commit is a major refactor of incidence2 that we are undertaking for the 2.0.0 release. Whilst we are making many changes internally, the two big changes that will affect users are (1), the removal of support for non-standard evaluation (NSE); and (2), the removal of the date manipulation functionality.

Our motivation for removing support for NSE are both the complexity it can bring to the underlying code (making long term maintenance harder) and the complexity it can cause for other users / developers who want to build on top of incidence2.

TimTaylor committed:
refactor for 2.0.0

This commit is a major refactor of incidence2 that we are undertaking for
the 2.0.0 release. Whilst we are making many changes internally, the two
big changes that will affect users are (1), the removal of support for
non-standard evaluation (NSE); and (2), the removal of the date
manipulation functionality.

Our motivation for removing support for NSE are both the complexity it can
bring to the underlying code (making long term maintenance harder) and the
complexity it can cause for other users / developers who want to build on
top of incidence2.

TimTaylor commented:
@thibautjombart - Let me know what you think. It is a very large refactor. You will need the development version of grates (main branch). Best to build the pkgdown site locally to get a feel for things. Main outstanding task for me is to think about #88 and then get on to fixing the dependencies (mainly i2extras) for compatibility.

thibautjombart commented:
Thanks for this, I will get to it this week.

thibautjombart commented on vignettes/handling_incidence_objects.Rmd:
```diff
@@ -17,88 +17,121 @@ knitr::opts_chunk$set(
 ### `keep_first()` and `keep_last()`
-Once your data is grouped by date, you may want to select the first or last few entries based on a particular date grouping using `keep_first()` and `keep_last()`:
+Once your data is grouped by date, you may want to select the first or last few
+entries based on a particular date grouping using `keep_first()` and
```
'entries' is ambiguous, as it could mean the last non-zero data points; should it read the last 'dates'?

TimTaylor replied:
This is a little tricky and also relates to #94. See discussion there.

thibautjombart commented on vignettes/incidence2.Rmd:
```diff
@@ -0,0 +1,137 @@
+  last) set of grouped dates (ordered by time) from an `<incidence>` object.
+  
+- `complete_counts()`: ensure every possible combination of date and groupings
+  is given an explicit count.
```
```suggestion
  is given an explicit count, adding zeros as needed.
```

TimTaylor replied:
Makes sense. Will leave this open for moment and may do grammatical changes all in one commit.

thibautjombart commented on vignettes/incidence2.Rmd:
````diff
@@ -0,0 +1,137 @@
+we can use to index the input:
+
+```{r daily, fig.height = 5, dpi = 90}
+(daily <- incidence(dat, date_index = "date_of_onset"))
````
Are the outside parentheses needed?

thibautjombart replied:
Unsure about this so will flag the rest as suggested changes.

TimTaylor replied:
prints, the result without having to retype `daily`. I find it neater but happy either way.

thibautjombart replied:
I think it should boild down to what most users are familiar with. I am quite possiby just old-fashioned, but I have not seen it used as a new default (I think?), so maybe best just type the name of the object we want to print?

TimTaylor replied:
Moving this to an issue.

thibautjombart commented on vignettes/incidence2.Rmd:
````diff
@@ -0,0 +1,137 @@
+we can use to index the input:
+
+```{r daily, fig.height = 5, dpi = 90}
+(daily <- incidence(dat, date_index = "date_of_onset"))
````
```suggestion
daily <- incidence(dat, date_index = "date_of_onset")
```

thibautjombart commented on vignettes/incidence2.Rmd:
````diff
@@ -0,0 +1,137 @@
+```{r sevenday}
+# 7 day incidence
+dat <- transform(dat, date_of_onset = as_period(date_of_onset, n = 7L))
+(inci <- incidence(dat, date_index = "date_of_onset"))
````
```suggestion
inci <- incidence(dat, date_index = "date_of_onset")
```

thibautjombart commented on vignettes/incidence2.Rmd:
````diff
@@ -0,0 +1,137 @@
+`groups` argument. For instance, we can compute incidence by gender:
+
+```{r gender}
+(grouped_inci <- incidence(
+    dat,
+    date_index = "date_of_onset",
+    groups = "gender"
+))
````
```suggestion
grouped_inci <- incidence(
    dat,
    date_index = "date_of_onset",
    groups = "gender"
)
```

thibautjombart commented on vignettes/incidence2.Rmd:
````diff
@@ -0,0 +1,137 @@
+Similarly, `incidence()` also supports multiple date inputs where the different
+date indices are used for filling the resultant plot
+```{r}
+dat <- transform(dat, date_of_infection = as_period(date_of_infection, n = 7L))
+(grouped_inci <- incidence(
+    dat,
+    date_index = c("date_of_onset", "date_of_infection")
+))
````
```suggestion
dat <- transform(dat, date_of_infection = as_period(date_of_infection, n = 7L))
grouped_inci <- incidence(
    dat,
    date_index = c("date_of_onset", "date_of_infection")
)
```

thibautjombart reviewed (commented):
I looked through the code but in the process of reinstalling R on my linux system to check the new version locally - will get this finished next week. 

My 2-cents at this stage: 

* this is tons of work, so a **big thanks** in the meantime - awesome stuff
* great to remove the NSE
* the only code idiom remaining seems to relate to `data.table`, but there are only few instances, so I think this is a big step forward in terms of ease of maintenance
* I will test this new version once I have a working, up-to-date, system

thibautjombart commented:
Note: I have created a project to store a backlog of tweaks to make once the version2 has been merged, mostly things I have noticed while reviewing the PR which are not deal-breakers and should not hold off the merging, but could be useful to do before the release.

thibautjombart reviewed (commented):
## Overview 

Thanks again for all the changes and hard work! I can see two types of changes to discuss:

* some are improvements which can be addressed after merging, possibly before the release; they are listed in [this project](https://github.com/orgs/reconverse/projects/1/views/1)
* some are more essential and are listed below for further discussion

## Things to discuss

The common thread here is mostly trying to preserve backward compatibility as long as the current design allows it without major changes. Here is a list of things we could discuss:

* bringing back the `interval` argument in `incidence`; it provides a major convenience for generating different epicurves quickly, and would be simpler than having users resort to `as_period` (even if under the hood, this is what happens)

* bringing back the `facet_plot`: I like the current implementation better, but for compatibility purposes we could have this function as a wrapper around `plot` with a deprecation message

* fixing `keep_first()` and `keep_last()`'s behaviour - see [corresponding issue](https://github.com/reconverse/incidence2/issues/94)

* plots: discuss how different date indices and groups are handled, including:
    + removing default color when there is but on date index and no group
    + avoiding the default to `dodge` for x positioning as it creates an illusion of delay
    + defaulting to facets for different dates indices as the y-scale could be widely different (e.g. reported cases vs deaths)
    + bringing back the default y-axis labels 

* the use of `na_as_zero`: it seems to me we should only ever get NAs in counts for a given time period / group from pre-aggregated counts clearly indicating `NA` in some entries; if so, these NAs should be kept as is; any other NAs occuring from completing combinations should be zeros (same rationale as in `complete_counts`); maybe we could remove the argument?

TimTaylor committed:
remove na_as_zero parameter from incidence

TimTaylor commented:
`na_as_zero` parameter now removed via https://github.com/reconverse/incidence2/pull/91/commits/9fd7c82f0b9a038f1f07c786fa97a7fec5bb6736 . It actually had less scope then intended after the refactor anyway so sensible to remove.

TimTaylor committed:
bump coverage action

TimTaylor committed:
wip - dplyr integration

TimTaylor committed:
wip - simplify incidence() and continue dplyr integration

TimTaylor committed:
wip - dplyr integration - expand tbl printing commentary

TimTaylor committed:
wip - dplyr integration - complete_counts

TimTaylor committed:
wip - dplyr integration - cumulate

TimTaylor committed:
remove hardcoding and simplify incidence implementation

TimTaylor committed:
remove testing hangover

TimTaylor committed:
provide a non data.table check for duplicate rows

Added a use of vctrs::vec_duplicate_any within the subsetting code for when
list columns are present.

We could only use the vctrs::vec_duplicate_any function but I want to
explicitly branch on non-data.table use as it will allow us to easily
drop dplyr support should we ever wish to do so.

TimTaylor committed:
turn on the dplyr support for non-atomic vectors

TimTaylor committed:
make checks safe re suggested packages

TimTaylor committed:
repermit relabelling of date indices during construction

TimTaylor committed:
More sensible plotting defaults

TimTaylor commented:
I've now implemented more appropriate default plots:

* Multiple counts will always now appear on seperate rows.
* Multiple counts and or multiple groups always appear as different facets.

TimTaylor committed:
FIX: ensure data is copied before using setnames!

TimTaylor committed:
plotting refinements

TimTaylor committed:
vignette tweaks

TimTaylor committed:
move {grates} form imports to depends

TimTaylor committed:
only use data.table

TimTaylor committed:
wip - allow use of interval

TimTaylor committed:
wip continued - allow use of interval

TimTaylor committed:
rename complete_counts() to complete_dates()

TimTaylor committed:
reintroduce error on POSIXlt or vctrs_rcrd columns

TimTaylor committed:
keep funs now optionally calls complete_dates (TRUE by default)

TimTaylor committed:
fix switch usage

TimTaylor committed:
revert maintainer change

TimTaylor committed:
update vignette with  examples

TimTaylor committed:
update NEWS

TimTaylor committed:
Add the defunct documentation

TimTaylor commented:
Closing this. I'm going to squash and then push to main. @thibautjombart - I've addressed most of what we've discussed and we utilise issues to finalise prior to release.
